src/overlords/situational/stronghold.ts

```diff
@@ -0,0 +1,316 @@
+	 * @param range
+	 * @param myCreep
+	 */
+	private findAttackingPositionAndTarget(target: Creep | Structure, range: number,
+										   myCreep: CombatZerg): {attackPos: RoomPosition, target: Creep | Structure} | undefined {
+		log.info(`Finding attacking position in ${target.room} for ${this.print}`);
+		if (!target.room || range == 0) {return;}
+
+		// bug with containers
+		const shootPositions = target.pos.getPositionsAtRange(range, false, false);
+
+		// Index is range to avoid from
+		const avoidLocs: RoomPosition[][] = Array.from({length: 5}, () => []);
+
+		avoidLocs[1] = avoidLocs[1].concat(target.room.ramparts.map(rampart => rampart.pos));
+		avoidLocs[4] = avoidLocs[4].concat(target.room.sources.map(s => s.pos));
+		avoidLocs[4] = avoidLocs[4].concat(target.room.keeperLairs.map(s => s.pos));
+		if (target.room.mineral) {avoidLocs[4] = avoidLocs[4].concat(target.room.mineral.pos);}
+
+		// Array where first index is how many hostile locations there are
+		const safeSpots: RoomPosition[][] = this.findSafeLocation(shootPositions, avoidLocs);
+		_.forEach(safeSpots, (distanceArray, hostilesIndex) => distanceArray.forEach(
+			spot => Visualizer.marker(spot, {color: StrongholdOverlord.numberToColor(hostilesIndex), frames: 2})));
+
+		// If you can safely attack goal, do that
+		if (safeSpots[0].length > 0) {
+			const closestFirst = safeSpots[0].sort(((a, b) =>
+				this.bestRampartToAttackSortFunction(a, b, myCreep.pos)));
+			return {attackPos: closestFirst[0], target: target};
+		} else if (safeSpots[1].length > 0) {
+			// Can't safely attack target, need to attack another position first
+			const range1Spots = safeSpots[1];
+			const ramparts = target.room!.ramparts;
+			const posToRampartMap: Map<RoomPosition, StructureRampart[]> = new Map();
+			for (const spot of range1Spots) {
+				for (const rampart of ramparts) {
+					if (rampart.pos.isNearTo(spot)) {
+						const temp: StructureRampart[] | undefined = posToRampartMap.get(spot);
+						posToRampartMap.set(spot, !!temp ? temp.concat([rampart]): [rampart]);
+					}
+				}
+			}
+
+			// Now select closest
+			if (myCreep) {
+				// TODO currently extra filtering, don't need from above
+				const orderedByBest = Array.from(posToRampartMap.keys()).filter(p => posToRampartMap.get(p)
+					&& posToRampartMap.get(p)!.length == 1).sort((a,b) => this.bestRampartToAttackSortFunction(a, b, myCreep.pos));
+				console.log('sorted array is ' + orderedByBest);
+				for (const pos of orderedByBest) {
+					const res = this.findAttackingPositionAndTarget(posToRampartMap.get(pos)![0], range-1, myCreep);
+					if (res) {
+						return res;
+					}
+				}
+			} else {
+				for (const pos of posToRampartMap) {
+					const res = this.findAttackingPositionAndTarget(pos[1][0]!, range-1, myCreep);
+					if (res) {
+						return res;
+					}
+				}
+			}
+		}
+		return;
+	}
+
+	bestRampartToAttackSortFunction(element1: RoomPosition, element2: RoomPosition, currentPos: RoomPosition,
```

Comment:
this is only used once and the function declaration is longer than the actual logic itself -  just make an anonymous function